Approved/enhanced-array-literals.md

```diff
@@ -0,0 +1,209 @@
+
+```qsharp
+use qs = Qubit[n];
+```
+
+For consistency, it may make sense to change this syntax as well.
+The [proposal for allocatable types and generalization of initializers](https://github.com/microsoft/qsharp-language/pull/41) is addressing this.
+
+## Impact on Existing Mechanisms
+
+The [`Default`](https://docs.microsoft.com/en-us/qsharp/api/qsharp/microsoft.quantum.core.default) function depends on default-initialized array constructors, and it will also be deprecated.
+
+## Anticipated Interactions with Future Modifications
+
+### Named arguments
+
+The syntax for repeated constant arrays is designed to be similar to the syntax for named arguments, a feature that Q# does not currently support (and for which there is currently no suggestion or proposal).
+This proposal makes the assumption that the syntax for named arguments, if Q# supports them, will be:
+
+```qsharp
+Foo(firstArgument, secondArgument, namedArgument = "foo");
+```
+
+### Type inference
+
+In the future, the type inference used for empty array literals may be extended to all expressions, allowing calls to polymorphic functions where not all type variables are determined by their arguments.
+
+### Partial application
+
+We may want to consider supporting partial application for array literals in the future:
+
+```qsharp
+let f = ["foo", size = _];
+// f : Int -> String[]
+// f(3) == ["foo", "foo", "foo"]
+
+let g = [_, size = 3];
+// g : 'a -> 'a[]
+// g("foo") == ["foo", "foo", "foo"]
+
+let h = [_, "middle", _];
+// h : (String, String) -> String[]
+// h("first", "last") == ["first", "middle", "last"]
+```
+
+## Alternatives
+
+Both empty arrays and repeated constant arrays can be created using functions in the Q# standard library, using the `EmptyArray` and `ConstantArray` functions.
+After deprecating the existing default-initialized array constructor syntax, this proposal could have made these functions the primary way to create these kinds of arrays instead of introducing new syntax.
+
+For empty arrays, the syntax `[]` comes naturally as the zero-length case of general array literals, and is very common in other languages.
+It is much more convenient to use than `EmptyArray<T>()`.
+
+For repeated constant arrays, the syntax `[x, size = n]` is not as obvious.
+Many languages use a function to provide this functionality instead of built-in syntax.
+However, we were not able to come to a consensus on how the `ConstantArray` function should be made available.
+There are unresolved questions, such as:
+
+1. Should `ConstantArray` be added to `Microsoft.Quantum.Core` so that it can be used without an open directive?
+2. Should `ConstantArray` be part of a more general `Array` module that is in `Microsoft.Quantum.Core`, so that it can be used like `Array.Constant`?
+3. Should modules be a language feature in Q# or are open-as directives enough to emulate them?
+
+We defer answering these questions for now, instead adding new syntax for this case.
```

Comment:
Saying we added new syntax because we couldn't come to a consensus is accurate. Given the inconveniences and downsides of each possible ways in which we could expose this via a library function and given how common array constructions are, it is justified to introduce new syntax. You can list the downsides of the exposing it via a library (copy-pasting from the rather long thread), or you can leave it at that.

Reply:
I disagree that the new syntax is justified when compared with an alternative function API that we could agree on. The thread is very long, but I don't remember any downsides of using a function other than needing to decide what to call it and where to put it. I remember some arguments about wanting to avoid duplicating `ConstantArray`, `ConstantSeq`, etc. if Q# has multiple collection types, but these same arguments apply to `[x, size = n]`, which is also array-specific.

Reply:
There are downsides to every possibility that are discussed in detail in the thread. I understand your assessment is that there are significant downsides to introducing new syntax that outweight the benefits of providing an alternative to a library call (which you can also use). I very much appreciate your thoughts and discussions about it with you (hence the length of the thread above). Given that we take care that the introduced syntax does not negatively impact future features and ties in well with anticipated features, I believe the best path forward is nonetheless to provide this alternative.